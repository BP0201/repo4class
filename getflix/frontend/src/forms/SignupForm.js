import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "../other/Alert";
import BackendAPI from "../apis/BackendApi";

function SignupForm({ signup }) {
  const navigate = useNavigate();
  const [formData, setFormData] = useState({
    username: "",
    password: ""
    // firstName: "",
    // lastName: "",
    // email: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  // console.debug(
  //     "SignupForm",
  //     "signup=", typeof signup,
  //     "formData=", formData,
  //     "formErrors=", formErrors,
  // );

  // handle form submission
  async function handleSubmit(evt) {
    evt.preventDefault();
    if (formData.username.length < 3) {
      setFormErrors(formErrors => [...formErrors, "Username must be between 3-16 characters"])
    }
    if (formData.password.length < 5) {
      setFormErrors(formErrors => [...formErrors, "Password must be between 5-20 characters"])
    }
    const user = await BackendAPI.getCurrentUser(formData.username);
    console.log("User:", user)
    if (!(!user)) {
      // check if user already exists
      setFormErrors(formErrors => [...formErrors, "Username taken"])
    }
    const result = await signup(formData);
    if (result.success) {
      // redirect if successful
      navigate("/search");
    }
  }

  // handle input value change
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({ ...data, [name]: value }));
  }

  return (
      <div className="SignupForm">
        <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
          <h2 className="my-4">Sign Up</h2>
          <div className="card shadow">
            <div className="card-body">
              <form onSubmit={handleSubmit}>
                <div className="form-group">
                  <label htmlFor="username">Username</label>
                  <input
                      id="username"
                      name="username"
                      className="form-control"
                      value={formData.username}
                      onChange={handleChange}
                      minLength={3}
                      maxLength={16}
                  />
                </div>
                <div className="form-group">
                  <label htmlFor="password">Password</label>
                  <input
                      id="password"
                      type="password"
                      name="password"
                      className="form-control"
                      value={formData.password}
                      onChange={handleChange}
                      minLength={5}
                      maxLength={20}
                  />
                </div>

                {/* <div className="form-group">
                  <label>First name</label>
                  <input
                      name="firstName"
                      className="form-control"
                      value={formData.firstName}
                      onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <label>Last name</label>
                  <input
                      name="lastName"
                      className="form-control"
                      value={formData.lastName}
                      onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <label>Email</label>
                  <input
                      type="email"
                      name="email"
                      className="form-control"
                      value={formData.email}
                      onChange={handleChange}
                  />
                </div> */}

                {formErrors.length
                    ? <Alert type="danger" messages={formErrors} />
                    : null
                }

                <button
                    data-testid="signup-form-btn"
                    type="submit"
                    className="btn btn-success float-right"
                    onSubmit={handleSubmit}
                >
                  Sign up
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
  );
}

export default SignupForm;