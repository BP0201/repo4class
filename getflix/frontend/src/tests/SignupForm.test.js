import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import SignupForm from '../forms/SignupForm';

describe("SignupForm", () => {
  it("should match the snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <SignupForm />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByLabelText("Username")).toBeInTheDocument();
    expect(screen.getByLabelText("Password")).toBeInTheDocument();
  });

  it("should throw an error if inputs are empty", () => {
    render(
      <MemoryRouter>
        <SignupForm />
      </MemoryRouter>
    );
    const signupBtn = screen.getByTestId("signup-form-btn");
    fireEvent.click(signupBtn);
    expect(screen.getByText("Username must be between 3-16 characters")).toBeInTheDocument();
    expect(screen.getByText("Password must be between 5-20 characters")).toBeInTheDocument();
  });
});