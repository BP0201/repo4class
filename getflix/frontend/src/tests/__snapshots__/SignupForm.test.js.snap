// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignupForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h2
        class="my-4"
      >
        Sign Up
      </h2>
      <div
        class="card shadow"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="form-group"
            >
              <label
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                maxlength="16"
                minlength="3"
                name="username"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                maxlength="20"
                minlength="5"
                name="password"
                type="password"
                value=""
              />
            </div>
            <button
              class="btn btn-success float-right"
              data-testid="signup-form-btn"
              type="submit"
            >
              Sign up
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
